lib/install/management/configure.go

```diff
@@ -98,50 +98,63 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
 
 	snapshotName := fmt.Sprintf("%s %s", ConfigurePrefix, conf.Version.BuildNumber)
 	snapshotName = strings.TrimSpace(snapshotName)
-
 	// check for old snapshot
-	// #nosec: Errors unhandled.
-	oldSnapshot, _ := d.appliance.GetCurrentSnapshotTree(d.op)
+	oldSnapshot, err := d.appliance.GetCurrentSnapshotTree(d.op)
+	if err != nil {
+		// log the error but continue
+		d.op.Debugf("Error checking appliance snapshot tree during %s: %s", d.Action.String(), err.Error())
+	}
 
 	newSnapshotRef, err := d.createSnapshot(snapshotName, "configure snapshot")
 	if err != nil {
-		d.deleteUpgradeImages(ds, settings)
+		d.deleteISOs(ds, settings)
 		return err
 	}
 
-	err = d.update(conf, settings, isConfigureOp)
+	// rollback function
+	rollback := func() {
+		// Roll back
+		d.op.Errorf("Failed to %s: %s", d.Action.String(), err)
+		d.op.Infof("Rolling back %s", d.Action.String())
 
-	// If successful try to grant permissions to the ops-user
-	if err == nil && conf.ShouldGrantPerms() {
-		err = opsuser.GrantOpsUserPerms(d.op, d.session, conf)
-		if err != nil {
-			// Update error message and fall through to roll back
-			err = errors.Errorf("Failed to grant permissions to ops-user, failure: %s", err)
+		if rerr := d.rollback(conf, snapshotName, settings); rerr != nil {
+			d.op.Errorf("Failed to revert appliance to snapshot: %s", rerr)
+			return
 		}
+		d.op.Infof("Appliance is rolled back to previous version")
+		d.deleteISOs(ds, settings)
+		d.deleteSnapshot(newSnapshotRef, snapshotName, conf.Name)
 	}
 
+	err = d.update(conf, settings)
 	if err != nil {
-		// Roll back
-		d.op.Errorf("Failed to upgrade: %s", err)
-		d.op.Infof("Rolling back upgrade")
+		rollback()
+		return err
+	}
 
-		if rerr := d.rollback(conf, snapshotName, settings); rerr != nil {
-			d.op.Errorf("Failed to revert appliance to snapshot: %s", rerr)
+	// if we are upgrading evaluate need for inventory upgrade
+	// vApp support planned: https://github.com/vmware/vic/issues/7670
+	if d.Action == UpgradeAction && d.session.IsVC() && d.vchPool.Reference().Type != "VirtualApp" {
+		err := d.inventoryUpdate(conf.Name)
+		if err != nil {
+			rollback()
 			return err
 		}
-		d.op.Infof("Appliance is rolled back to old version")
-
-		d.deleteUpgradeImages(ds, settings)
-		d.retryDeleteSnapshotByRef(newSnapshotRef, snapshotName, conf.Name)
+	}
 
-		// return the error message for upgrade
-		return err
+	// If successful try to grant permissions to the ops-user
+	if conf.ShouldGrantPerms() {
+		err := opsuser.GrantOpsUserPerms(d.op, d.session, conf)
+		if err != nil {
+			err = errors.Errorf("Failed to grant permissions to ops-user, failure: %s", err)
```

Comment:
Should this be `return errors.Errorf(...)`? Otherwise it seems to be ignored.

Reply:
hmmm...that seems problematic...i.e. needs to be changed...